From 135ff349b7bf38d92b0b103e4865474b75d860fd Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Mon, 12 Oct 2026 01:43:32 +0200
Subject: [PATCH] MFront: static variables given to the material properties in
 device code

The code generated by MFront wrote `Class::v` for a static variable used
as an input of a material property (elastic material properties, thermal
expansion coefficients, material properties of the implicit, isotropic,
Runge-Kutta and default DSLs, bricks, DDIF2 stress potential), given by
reference in device code for instance to base_type_cast. It is now
written `TFEL_STATIC_VARIABLE(Class::v)`, as in the code blocks: the
variable itself in host code, a copy in device code.
---
 mfront/src/BehaviourCodeGeneratorBase.cxx          |  5 +++--
 mfront/src/BrickUtilities.cxx                      |  3 ++-
 mfront/src/DDIF2StressPotential.cxx                |  2 +-
 mfront/src/DefaultCodeGeneratorBase.cxx            |  3 ++-
 mfront/src/ImplicitCodeGeneratorBase.cxx           | 12 ++++++++----
 mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx |  9 ++++++---
 mfront/src/RungeKuttaCodeGeneratorBase.cxx         |  3 ++-
 7 files changed, 24 insertions(+), 13 deletions(-)

diff --git a/mfront/src/BehaviourCodeGeneratorBase.cxx b/mfront/src/BehaviourCodeGeneratorBase.cxx
index 0a499faad..fe885d9a4 100644
--- a/mfront/src/BehaviourCodeGeneratorBase.cxx
+++ b/mfront/src/BehaviourCodeGeneratorBase.cxx
@@ -2880,8 +2880,9 @@ namespace mfront {
               {i.name, "tfel::math::base_type_cast(this->" + i.name + ")"});
         } else if (c == BehaviourDescription::MaterialPropertyInput::
                             STATICVARIABLE) {
-          m.insert({i.name, "tfel::math::base_type_cast(" +
-                                this->bd.getClassName() + "::" + i.name + ")"});
+          m.insert({i.name, "tfel::math::base_type_cast(TFEL_STATIC_VARIABLE(" +
+                                this->bd.getClassName() + "::" + i.name +
+                                "))"});
         } else {
           throw_if(true,
                    "thermal expansion coefficients must depend "
diff --git a/mfront/src/BrickUtilities.cxx b/mfront/src/BrickUtilities.cxx
index 7cee8cc80..ace341ee4 100644
--- a/mfront/src/BrickUtilities.cxx
+++ b/mfront/src/BrickUtilities.cxx
@@ -37,7 +37,8 @@ namespace mfront::bbrick {
                  (i.category == MaterialPropertyInput::PARAMETER)) {
         return "this->" + i.name;
       } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-        return bd.getClassName() + "::" + i.name;
+        return "TFEL_STATIC_VARIABLE(" + bd.getClassName() + "::" + i.name +
+               ")";
       } else {
         tfel::raise("unsupported input type for variable '" + i.name + "'");
       }
diff --git a/mfront/src/DDIF2StressPotential.cxx b/mfront/src/DDIF2StressPotential.cxx
index a1a494efa..f2544f4cc 100644
--- a/mfront/src/DDIF2StressPotential.cxx
+++ b/mfront/src/DDIF2StressPotential.cxx
@@ -334,7 +334,7 @@ namespace mfront::bbrick {
       }
       throw_if(i.category != MaterialPropertyInput::STATICVARIABLE,
                "unsupported input type for variable '" + i.name + "'");
-      return bd.getClassName() + "::" + i.name;
+      return "TFEL_STATIC_VARIABLE(" + bd.getClassName() + "::" + i.name + ")";
     };
     const auto young = [&bd]() -> std::string {
       if (bd.areElasticMaterialPropertiesDefined()) {
diff --git a/mfront/src/DefaultCodeGeneratorBase.cxx b/mfront/src/DefaultCodeGeneratorBase.cxx
index 15df96fed..8594e0bf3 100644
--- a/mfront/src/DefaultCodeGeneratorBase.cxx
+++ b/mfront/src/DefaultCodeGeneratorBase.cxx
@@ -41,7 +41,8 @@ namespace mfront {
                  (i.category == MaterialPropertyInput::PARAMETER)) {
         return "this->" + i.name;
       } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-        return this->bd.getClassName() + "::" + i.name;
+        return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+               "::" + i.name + ")";
       } else {
         tfel::raise(
             "DefaultCodeGeneratorBase::"
diff --git a/mfront/src/ImplicitCodeGeneratorBase.cxx b/mfront/src/ImplicitCodeGeneratorBase.cxx
index 121240b74..e7b2ab015 100644
--- a/mfront/src/ImplicitCodeGeneratorBase.cxx
+++ b/mfront/src/ImplicitCodeGeneratorBase.cxx
@@ -169,7 +169,8 @@ namespace mfront {
                  (i.category == MaterialPropertyInput::PARAMETER)) {
         return "this->" + i.name;
       } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-        return this->bd.getClassName() + "::" + i.name;
+        return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+               "::" + i.name + ")";
       } else {
         this->throwRuntimeError(
             "ImplicitCodeGeneratorBase::"
@@ -188,7 +189,8 @@ namespace mfront {
                  (i.category == MaterialPropertyInput::PARAMETER)) {
         return "this->" + i.name;
       } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-        return this->bd.getClassName() + "::" + i.name;
+        return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+               "::" + i.name + ")";
       } else {
         this->throwRuntimeError(
             "ImplicitCodeGeneratorBase::"
@@ -318,7 +320,8 @@ namespace mfront {
                  (i.category == MaterialPropertyInput::PARAMETER)) {
         return "this->" + i.name;
       } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-        return this->bd.getClassName() + "::" + i.name;
+        return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+               "::" + i.name + ")";
       } else {
         this->throwRuntimeError(
             "ImplicitCodeGeneratorBase::"
@@ -337,7 +340,8 @@ namespace mfront {
                  (i.category == MaterialPropertyInput::PARAMETER)) {
         return "this->" + i.name;
       } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-        return this->bd.getClassName() + "::" + i.name;
+        return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+               "::" + i.name + ")";
       } else {
         this->throwRuntimeError(
             "ImplicitCodeGeneratorBase::"
diff --git a/mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx b/mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx
index c86bb1fc0..ac52d57f9 100644
--- a/mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx
+++ b/mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx
@@ -38,7 +38,8 @@ namespace mfront {
                  (i.category == MaterialPropertyInput::PARAMETER)) {
         return "this->" + i.name;
       } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-        return this->bd.getClassName() + "::" + i.name;
+        return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+               "::" + i.name + ")";
       } else {
         this->throwRuntimeError(
             "IsotropicBehaviourCodeGeneratorBase::"
@@ -56,7 +57,8 @@ namespace mfront {
                  (i.category == MaterialPropertyInput::PARAMETER)) {
         return "this->" + i.name;
       } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-        return this->bd.getClassName() + "::" + i.name;
+        return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+               "::" + i.name + ")";
       } else {
         this->throwRuntimeError(
             "IsotropicBehaviourCodeGeneratorBase::"
@@ -204,7 +206,8 @@ namespace mfront {
               (i.category == MaterialPropertyInput::PARAMETER)) {
             return "this->" + i.name;
           } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-            return this->bd.getClassName() + "::" + i.name;
+            return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+                   "::" + i.name + ")";
           } else {
             this->throwRuntimeError(
                 "IsotropicMisesCreepCodeGenerator::writeBehaviourIntegrator",
diff --git a/mfront/src/RungeKuttaCodeGeneratorBase.cxx b/mfront/src/RungeKuttaCodeGeneratorBase.cxx
index 0fae3dfdf..e140279d7 100644
--- a/mfront/src/RungeKuttaCodeGeneratorBase.cxx
+++ b/mfront/src/RungeKuttaCodeGeneratorBase.cxx
@@ -295,7 +295,8 @@ namespace mfront {
             (i.category == MaterialPropertyInput::PARAMETER)) {
           return "this->" + i.name;
         } else if (i.category == MaterialPropertyInput::STATICVARIABLE) {
-          return this->bd.getClassName() + "::" + i.name;
+          return "TFEL_STATIC_VARIABLE(" + this->bd.getClassName() +
+                 "::" + i.name + ")";
         } else {
           this->throwRuntimeError(
               "RungeKuttaCodeGeneratorBase::"
-- 
2.43.0

